solutions/platinum/csa-RootLCAQueries.mdx

```diff
@@ -0,0 +1,155 @@
+---
+id: csa-RootLCAQueries
+source: CS Academy
+title: Root LCA Queries
+author: Justin Ji
+---
+
+<Spoiler title = "Hint 1">
+
+Let $a$, $b$, and $c$ be the three nodes given in each query.
+When will the number of possible roots be greater than zero?
+
+</Spoiler>
+
+<Spoiler title = "Answer to Hint 1">
+
+The answer for a given query is only greater than zero when $c$ is on the path
+from $a$ to $b$, as by definition the LCA must lie on the path from $a$ to $b$.
+Now try doing a bit of casework to calculate the answer.
+
+</Spoiler>
+
+<Spoiler title = "Solution">
+
+## Explanation
+
+As noted in the hints, the LCA of $a$ and $b$ must lie on the path between these
+nodes. Arbitrarily root the tree, then consider the following cases:
+
+**Case 1: $c$ is not on the path from $a$ to $b$.**
+
+The answer is $0$ for this case.
+
+**Case 2: $c$ is equal to the LCA of $a$ and $b$, for our chosen root.**
+
+Any node outside of the subtree of $c$ is a suitable root. Within the subtree of
+$c$, any node whose path to node $c$ does not intersect the path from $a$ to $b$
+is a suitable root.
+
+**Case 3: $c$ is on the path from $a$ to $b$, but is not equal to the LCA of
+$a$ and $b$.**
+
+This case is handled the same way as case 2, except only nodes in the subtree
+of $c$ can be possible roots.
+
+To quickly answer queries, we use binary lifting to calculate the LCA, and
+precalculate the number of nodes in each subtree to quickly remove unsuitable
+subtrees from our answer.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N\log{N} + Q\log{N})$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+class Tree {
+  private:
+	const int n, log2dist;
+	const vector<vector<int>> &adj;
+	vector<vector<int>> lift;
+	vector<int> dep, sub;
+
+	/** calculates lift, dep, and sub arrays */
+	void dfs(int u, int p) {
+		for (int v : adj[u]) {
+			if (v == p) continue;
+			dep[v] = dep[u] + 1;
+			lift[0][v] = u;
+			for (int p = 1; p < log2dist; p++) {
+				lift[p][v] = lift[p - 1][lift[p - 1][v]];
+			}
+			dfs(v, u);
+			sub[u] += sub[v];
+		}
+	}
+
+	/** @return d-th ancestor of u */
+	int jump(int u, int d) {
+		for (int i = log2dist - 1; i >= 0; i--) {
+			if ((d >> i) & 1) { u = lift[i][u]; }
+		}
+		return u;
+	}
+
+	/** @return LCA of nodes u and v */
+	int lca(int u, int v) {
```

Comment:
i'd argue that these two methods could be made public but do what you want i spose

Reply:
uh if you're going to put this as resolved at least leave a remark :pleading_face:

Reply:
oops my bad g

i just put put it as private b/c u aren't gonna need to call it

Reply:
altho now that i think about it i might as well just do what u said, will do that rq